File: ios/Packages/FeatureServices/PaymentService/PaymentService.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import class Gemstone.GemPaymentService
import protocol Gemstone.GemPaymentServiceProtocol
import struct Gemstone.GemWalletConnectPayAuth
import GemstonePrimitives
import NativeProviderService
import Primitives

public protocol PaymentServiceable: Sendable {
    func getOptions(link: PaymentLink, addresses: [ChainAddress]) async throws -> PaymentOptions
    func getQuoteData(quote: PaymentQuote, addresses: [ChainAddress]) async throws -> PaymentQuoteData
    func confirm(payment: PaymentData, transactionHash: String) async throws -> PaymentOutcome
}

public final class PaymentService: Sendable {
    private let service: any GemPaymentServiceProtocol

    public init(service: any GemPaymentServiceProtocol) {
        self.service = service
    }

    public convenience init(
        nodeProvider: any NodeURLFetchable,
        auth: GemWalletConnectPayAuth,
        requestInterceptor: any RequestInterceptable = EmptyRequestInterceptor(),
    ) {
        self.init(
            service: GemPaymentService(
                provider: NativeProvider(nodeProvider: nodeProvider, requestInterceptor: requestInterceptor),
                walletConnectPay: auth,
            ),
        )
    }
}

// MARK: - PaymentServiceable

extension PaymentService: PaymentServiceable {
    public func getOptions(link: PaymentLink, addresses: [ChainAddress]) async throws -> PaymentOptions {
        try await service.getOptions(link: link.map(), addresses: addresses.map { $0.map() }).map()
    }

    public func getQuoteData(quote: PaymentQuote, addresses: [ChainAddress]) async throws -> PaymentQuoteData {
        try await service.getQuoteData(quote: quote.map(), addresses: addresses.map { $0.map() }).map()
    }

    public func confirm(payment: PaymentData, transactionHash: String) async throws -> PaymentOutcome {
        try await service.confirm(quote: payment.quote.map(), transactionHash: transactionHash).map()
    }
}

File: android/blockchain/src/main/kotlin/com/gemwallet/android/blockchain/services/PaymentService.kt
package com.gemwallet.android.blockchain.services

import com.gemwallet.android.ext.gemChainAddresses
import com.gemwallet.android.ext.toGem
import com.gemwallet.android.ext.toPrimitives
import com.wallet.core.primitives.PaymentLink
import com.wallet.core.primitives.PaymentOptions
import com.wallet.core.primitives.PaymentOutcome
import com.wallet.core.primitives.PaymentQuote
import com.wallet.core.primitives.PaymentQuoteData
import com.wallet.core.primitives.Wallet
import uniffi.gemstone.GemPaymentServiceInterface

class PaymentService(
    private val client: GemPaymentServiceInterface,
) {
    suspend fun getOptions(link: PaymentLink, wallet: Wallet): PaymentOptions =
        client.getOptions(link.toGem(), wallet.gemChainAddresses()).toPrimitives()

    suspend fun getQuoteData(quote: PaymentQuote, wallet: Wallet): PaymentQuoteData =
        requireNotNull(client.getQuoteData(quote.toGem(), wallet.gemChainAddresses()).toPrimitives()) {
            "Payment asks for an unsupported chain"
        }

    suspend fun confirm(quote: PaymentQuote, transactionHash: String): PaymentOutcome =
        client.confirm(quote.toGem(), transactionHash).toPrimitives()
}
